go/rows.go

```diff
@@ -105,39 +118,67 @@ func (r *Rows) Next(dest []driver.Value) error {
 	if r.reachedLastPage {
 		return io.EOF
 	}
-	if len(r.ResultOutput.ResultSet.Rows) == 0 {
-		if r.ResultOutput.NextToken == nil || *r.ResultOutput.NextToken == "" {
-			// this means we reach the last page - no token and no rows
-			r.reachedLastPage = true
-			return io.EOF
-		}
 
-		if err := r.fetchNextPage(r.ResultOutput.NextToken); err != nil {
-			return err
-		}
-		if r.reachedLastPage {
-			return io.EOF
-		}
+	next, err := r.results.Read()
+	if err != nil {
+		return err
 	}
 
-	// Shift to next row
-	cur := r.ResultOutput.ResultSet.Rows[0]
 	columns := r.ResultOutput.ResultSet.ResultSetMetadata.ColumnInfo
-	if err := r.convertRow(columns, cur.Data, dest, r.config); err != nil {
+	if err := r.convertRecord(columns, next, dest, r.config); err != nil {
 		return err
 	}
-	r.ResultOutput.ResultSet.Rows = r.ResultOutput.ResultSet.Rows[1:]
+
 	return nil
 }
 
-// fetchNextPage is to get next result set page with a specific token.
-func (r *Rows) fetchNextPage(token *string) error {
-	var err error
-	r.ResultOutput, err = r.athena.GetQueryResultsWithContext(r.ctx,
-		&athena.GetQueryResultsInput{
-			QueryExecutionId: aws.String(r.queryID),
-			NextToken:        token,
+func (r *Rows) openResults() error {
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
should we reuse the context from `NewRows` so that we can support cancelation properly?

Reply:
Didn't even notice it was stored; yes.